From 625aab02193c7af95edc4ba99dfd12d0b0b6b8c1 Mon Sep 17 00:00:00 2001
From: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
Date: Wed, 12 Mar 2025 13:46:25 +0100
Subject: [PATCH] libqcow2: prepare proper cleanup of libqcow2 on close

The close operation will stop the libqcow2 thread and we'll need to
clean-up resources to prepare the reopen or unpause.

Signed-off-by: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
---
 include/qemu/main-loop.h   |  2 ++
 qcow2/lib/system/cpus.c    |  5 ++++
 qcow2/lib/util/main-loop.c | 49 +++++++++++++++++++++++++++++++++++++-
 3 files changed, 55 insertions(+), 1 deletion(-)

diff --git a/include/qemu/main-loop.h b/include/qemu/main-loop.h
index feab3b96..c2983f59 100644
--- a/include/qemu/main-loop.h
+++ b/include/qemu/main-loop.h
@@ -42,6 +42,7 @@ typedef struct MainLoop MainLoop;
 #endif
 
 void qemu_init_cpu_loop(void);
+void qemu_deinit_cpu_loop(void);
 
 /**
  * qemu_init_main_loop: Set up the process so that it can run the main loop.
@@ -57,6 +58,7 @@ void qemu_init_cpu_loop(void);
  * In the case of QEMU tools, this will also start/initialize timers.
  */
 int qemu_init_main_loop(Error **errp);
+int qemu_deinit_main_loop(void);
 
 /**
  * main_loop_wait: Run one iteration of the main loop.
diff --git a/qcow2/lib/system/cpus.c b/qcow2/lib/system/cpus.c
index ed2b9ff5..dbf0b3dc 100644
--- a/qcow2/lib/system/cpus.c
+++ b/qcow2/lib/system/cpus.c
@@ -429,6 +429,11 @@ void qemu_init_cpu_loop(void)
     //qemu_thread_get_self(&io_thread);
 }
 
+void qemu_deinit_cpu_loop(void)
+{
+    qemu_mutex_destroy(&bql);
+}
+
 #if 0
 void run_on_cpu(CPUState *cpu, run_on_cpu_func func, run_on_cpu_data data)
 {
diff --git a/qcow2/lib/util/main-loop.c b/qcow2/lib/util/main-loop.c
index 817e65a6..8bde8add 100644
--- a/qcow2/lib/util/main-loop.c
+++ b/qcow2/lib/util/main-loop.c
@@ -39,8 +39,12 @@
 #include <sys/wait.h>
 #endif
 
+static void iohandler_deinit(void);
+
 #ifndef _WIN32
 
+static int sigfd = -1;
+
 /* If we have signalfd, we mask out the signals we want to handle and then
  * use signalfd to listen for them.  We rely on whatever the current signal
  * handler is to dispatch the signals when we receive them.
@@ -86,7 +90,6 @@ static void sigfd_handler(void *opaque)
 
 static int qemu_signal_init(Error **errp)
 {
-    int sigfd;
     sigset_t set;
 
     /*
@@ -191,6 +194,39 @@ int qemu_init_main_loop(Error **errp)
     return 0;
 }
 
+int qemu_deinit_main_loop(void)
+{
+    GSource *src;
+
+    src = iohandler_get_g_source();
+    g_source_unref(src);
+    g_source_destroy(src);
+    g_source_unref(src);
+
+    src = aio_get_g_source(qemu_aio_context);
+    g_source_unref(src);
+    g_source_destroy(src);
+    g_source_unref(src);
+
+    g_array_free(gpollfds, TRUE);
+
+    qemu_bh_delete(qemu_notify_bh);
+    qemu_notify_bh = NULL;
+
+    if (sigfd != -1) {
+        aio_set_fd_handler(iohandler_get_aio_context(), sigfd, NULL, NULL, NULL, NULL, NULL);
+        close(sigfd);
+        sigfd = -1;
+    }
+
+    iohandler_deinit();
+    qemu_aio_context = NULL;
+
+    timerlistgroup_deinit(&main_loop_tlg);
+
+    return 0;
+}
+
 #if 0
 static void main_loop_update_params(EventLoopBase *base, Error **errp)
 {
@@ -633,6 +669,17 @@ static void iohandler_init(void)
     }
 }
 
+static void iohandler_deinit(void)
+{
+    if (iohandler_ctx) {
+        /*
+         * iohandler_ctx will be freed by the callback aio_ctx_finalize,
+         * but we need to set it to null to reallocate a new one in the future.
+         */
+        iohandler_ctx = NULL;
+    }
+}
+
 AioContext *iohandler_get_aio_context(void)
 {
     iohandler_init();
